Add support for directory symlinks

hoxu opened the pull request:
Fixes #768.

hoxu commented:
@arcivanov could you have a look at this PR?

arcivanov commented:
I have looked over it preliminary, but I need to fix the Windows build first, to make sure your code works on all platforms.

hoxu commented:
@arcivanov is there an issue of the broken Windows build? Would someone be able to help fixing it?

I don't use Windows so I'm unfortunately not helpful here.

arcivanov commented:
I don't either, it is with the Windows environment and I have no time to fix it right now, unfortunately. I'll see if I can do something later today.

hoxu commented:
@arcivanov any updates on this? Is there an issue about the Windows build being broken? Could someone lend a hand fixing it?

hoxu commented:
Regardless, I think this change should be okay on Windows as well.

https://docs.python.org/3/library/os.html#os.walk

> By default, walk() will not walk down into symbolic links that resolve to directories. Set followlinks to True to visit directories pointed to by symlinks, on systems that support them.

Sounds like a no-op on platforms that don't support symlinks.

hoxu commented:
@arcivanov any updates regarding getting this rather trivial and supposedly safe PR merged?

Is there an issue about the broken Windows build, so that people could help with that?

arcivanov commented:
@hoxu could you please rebase on the latest master? Thanks!

arcivanov commented:
Now that we know that it builds, could you please setup integration tests to confirm that it functions appropriately?

hoxu commented:
Not sure how much time I can spare for this.

Do you have thoughts for testcases, and pointers to any similar tests?

arcivanov commented:
Let me see if I can add a test quickly.

arcivanov commented:
The integration test simply sets up a PyB project and then checks things for working. The actual harnesses are very powerful and well-developed so it may be just a few lines.

hoxu commented:
So basically just:

1. Create a file in `src/integrationtest/python/`.
1. Extend `IntegrationTestSupport`.
1. Call `self.write_build_file` with a simple template.
1. Create directory `src/main/python`.
1. Create directory `symlinked` with `something` inside it.
1. Symlink `src/main/python/symlinked` to `symlinked`.
1. Use `self.assert_file_content` to check that `something` ends up in the result.

Like https://github.com/pybuilder/pybuilder/blob/master/src/integrationtest/python/should_copy_resources_from_non_std_location_tests.py

Correct? Seems simple enough that maybe I could find time to cook it up, so let me know if you'd rather do it yourself or have me try.

arcivanov commented:
@hoxu yes, very much so!

hoxu commented:
I have some trouble getting the integration tests running locally.

I changed `build.py` she-bang to `#!/usr/bin/env python3` and tried to run integration tests with `./build.py run_integration_tests`. One test seems to fail:

```
[ERROR] Integration test failed: /Users/username/path/to/pybuilder/src/integrationtest/python/smoke_setup_tests.py, exit code 1
<clip>
BUILD FAILED - 1 of 41 integration tests failed. (pybuilder/plugins/python/test_plugin_helper.py:54)
------------------------------------------------------------
Build finished at 2021-09-07 09:50:33
Build took 347 seconds (347297 ms)
```

and for some reason my new testcase does not seem to be found.

So, here is a blind-coded draft: f2d6f6be78f395344a4483b2233adbfac58ae8ca

arcivanov commented:
In order for tests to be executed the file should be suffixed by `_tests.py`

arcivanov commented:
Please see this test I've just added for issue 770 for how to validate these issues using standard unittest instrumentation.

https://github.com/pybuilder/pybuilder/pull/785/files#diff-0868e0a95236bb39eff42bc6baff24ddf7a94e7a54a62baa095ededdde172626

hoxu commented:
Rebased on master, fixed the test by rewriting, commit 76e62fe50f2ded0feb5c7f426cb409afa168786e.

The problem is, this test is still not testing for right functionality, and it succeeds even without the preceding change.

I [described the problem](https://github.com/pybuilder/pybuilder/issues/768#issuecomment-777234157) earlier:

> Further digging revealed that symlinked directories are dereferenced properly, and copied to `target/dist/$name/`.
> 
> However, `setup.py -v clean --all bdist_wheel` does not include the copied files in the tarball, because they are not included in `packages` or `py_modules`.

So the resulting tarball should be tested, not the dist directory. I guess there are no such tests currently?

I guess another possibility would be writing a test for `list_packages`:

https://github.com/pybuilder/pybuilder/blob/32ec7c7b5cdf0e97d17d3dfb48e79408bb284088/src/main/python/pybuilder/plugins/python/core_plugin.py#L126-L133

I'm also afraid I'm running short on time I can spend on this 😔

arcivanov commented:
Ok, thanks, I'll take it on from here!

hoxu commented:
Thanks! If you build on this test, note that `cargo` should be renamed to `cargo.py` so that it's picked up by `list_packages` (with `followlinks=True`).

arcivanov commented:
Closing in favor of #786
